library/kotlin/test/io/envoyproxy/envoymobile/GRPCStreamEmitterTest.kt

```diff
@@ -0,0 +1,101 @@
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import java.io.ByteArrayOutputStream
+import java.nio.ByteBuffer
+
+class GRPCStreamEmitterTest {
+
+  // TODO: Problems with nhaarman/mockito-kotlin https://github.com/lyft/envoy-mobile/issues/504
+  // This is a total hack to get something to work
+  private lateinit var emitter: StreamEmitter
+
+  private val dataOutputStream = ByteArrayOutputStream()
+  private var isCloseCalled = false
+
+  @Before
+  fun setup() {
+    emitter = object : StreamEmitter {
+
+      override fun cancel() {
+        throw UnsupportedOperationException("unexpected usage of mock emitter")
+      }
+
+      override fun sendData(byteBuffer: ByteBuffer): StreamEmitter {
+        dataOutputStream.write(byteBuffer.array())
+        return this
+      }
+
+      override fun sendMetadata(metadata: Map<String, List<String>>): StreamEmitter {
+        throw UnsupportedOperationException("unexpected usage of mock emitter")
+      }
+
+      override fun close(trailers: Map<String, List<String>>?) {
+        isCloseCalled = true
+      }
+    }
+  }
+
+  @After
+  fun teardown() {
+    dataOutputStream.reset()
+    isCloseCalled = false
+  }
+
+  @Test
+  fun `prefix and data is sent on send data`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(dataOutputStream.toByteArray()).hasSize(payload.size + GRPC_PREFIX_LENGTH)
+  }
+
+  @Test
+  fun `compression flag is set on the first bit of the prefix`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(dataOutputStream.toByteArray()[0]).isEqualTo(0)
+  }
+
+  @Test
+  fun `message length is set on the 1-4 bytes of the prefix`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(ByteBuffer.wrap(dataOutputStream.toByteArray().sliceArray(1..4)).int).isEqualTo(payload.size)
+  }
+
+  @Test
+  fun `message is sent after the prefix`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    val payload = "data".toByteArray(Charsets.UTF_8)
+    val message = ByteBuffer.wrap(payload)
+    grpcStreamEmitter.sendMessage(message)
+
+    assertThat(dataOutputStream.toByteArray().sliceArray(GRPC_PREFIX_LENGTH until dataOutputStream.size()).toString(Charsets.UTF_8)).isEqualTo("data")
+
+  }
+
+  @Test
+  fun `close is called with empty data frame`() {
+    val grpcStreamEmitter = GRPCStreamEmitter(emitter)
+
+    grpcStreamEmitter.close()
+
+    assertThat(isCloseCalled).isTrue()
```

Comment:
This actually doesn't validate whether close is called with trailers or a data frame. We should check that trailers are `nil` in this case